Assets/_Scripts/GameStatus.cs

```diff
@@ -3,36 +3,142 @@
+		{
+			UserInputManager userCtrl = player.GetComponent<UserInputManager>();
+			int playerID = userCtrl.playerNum;
+			userDataCollection[playerID] = new UserData();
+			userDataCollection[playerID].userID = playerID;
+			userDataCollection[playerID].initPosition = player.transform.position;
+		}
 	}
 	
 	// Update is called once per frame
 	void Update () {
+
+	}
+
+	void WinEndGame()
+	{
 		GameObject[] playerCollection = GameObject.FindGameObjectsWithTag (TagList.Player);
-		if(playerCollection.Length <= 1)
+
+		isGameOver = true;
+
+		foreach(GameObject player in playerCollection)
 		{
-			if(playerCollection.Length == 1 && !isGameOver)
+			GameObject winEffPrefab = Resources.Load ("ArenaEffects/WinParEff") as GameObject;
+			GameObject winEff = GameObject.Instantiate (winEffPrefab, 
+			                                            player.transform.position, Quaternion.identity)
+				as GameObject;
+			UserInputManager userInput = player.GetComponent<UserInputManager>();
+			userInput.LockLeftInput(3.0f);
+		}
+		StartCoroutine(ReadyToRestart());
+	}
+
+	IEnumerator ReadyToRestart()
+	{
+		yield return new WaitForSeconds (3.0f);
+		Application.LoadLevel (Application.loadedLevel);
+	}
+
+	public void DecPlayerLife(int playerID)
+	{
+		playerDeathCount [playerID] ++;
+		DestroyPlayerWithID(playerID);
+
+		switch(gameMode)
+		{
+		case GameMode.LoseLife:
+			if(playerDeathCount [playerID] >= GameTargetRounds)
+			{
+
+				GameObject[] playerCollection = GameObject.FindGameObjectsWithTag (TagList.Player);
+
+				if(playerCollection.Length <= 2 && !isGameOver)
+				{
+					WinEndGame();
+				}
+			}
+			else
+			{
+				StartCoroutine(RebornPlayerWithID(playerID));
+			}
+			break;
+
+		case GameMode.GainScore:
+			int team = playerID >= 2 ? 0 : 2;
+			if(playerDeathCount [team] + playerDeathCount [team+1] >= GameTargetRounds)
+			{
+				DestroyPlayerWithID((team + 2)%4);
+				DestroyPlayerWithID((team + 3)%4);
+				WinEndGame();
+			}
+			else
 			{
-				isGameOver = true;
-				GameObject winEffPrefab = Resources.Load ("ArenaEffects/WinParEff") as GameObject;
-				GameObject winEff = GameObject.Instantiate (winEffPrefab, 
-						 	playerCollection[0].transform.position, Quaternion.identity)
-							as GameObject;
-				UserInputManager userInput = playerCollection[0].GetComponent<UserInputManager>();
-				userInput.LockLeftInput(3.0f);
-				StartCoroutine(ReadyToRestart());
+				StartCoroutine(RebornPlayerWithID(playerID));
 			}
+			break;
+		default:
+			break;
+
 		}
 
 	}
 
-	IEnumerator ReadyToRestart()
+	
+	IEnumerator RebornPlayerWithID(int id)
 	{
 		yield return new WaitForSeconds (3.0f);
-		Application.LoadLevel (Application.loadedLevel);
+		GameObject wizard = 
+			Instantiate(playerPrefab,userDataCollection [id].initPosition,Quaternion.identity) 
+				as GameObject;
+		UserInputManager userCtrl = wizard.GetComponent<UserInputManager>();
+		userCtrl.playerNum = id;
 	}
+
+	void DestroyPlayerWithID(int id)
```

Comment:
If you have something like this, why not use a hashtable 
to store the (id -> Player) mapping?